### ISSUE TITLE
clusterctl only considers 30 github releases

### ISSUE THREAD

**Issue by killianmuldoon** [kubernetes-sigs/cluster-api#8233]
When clusterctl looks for releases for providers it looks up the https://api.github.com/repos/kubernetes-sigs/cluster-api/releases endpoint. This endpoint by default only lists the most recent 30 releases. Even if an earlier release does exist clusterctl will not find it if it's not in the most recent 30 releases.

The current impact of this is that clusterctl v0.4.X and clusterctl v0.3.X no longer work to initialize providers. A workaround is to pull the provider yamls manually and put them in a local clusterctl repository.

v0.4.X and v0.3.X of Cluster API are out of date so there is no need to update those older versions to fix this bug. This should be fixed in supported versions of Cluster API as we don't control the number of releases done by providers and different approaches e.g. many published pre-releases, could result in an older supported version not being installable. 

The API call should be able to look through subsequent pages of results in order to find older valid tags. Note: This should also be checked for the gitlab API to see if it suffers from a similar limitation. 

/kind bug
/area clusterctl

**Comment by abhay-krishna** [comment]
I faced this issue when emulating the `ListReleases` call [here](https://github.com/kubernetes-sigs/cluster-api/tree/main/cmd/clusterctl/client/repository/repository_github.go#L287) and not the `GetReleaseByTag` call [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/client/cluster/template.go#L265) which does indeed fetch an older tagged release, as far back as v0.2.6. While testing using `curl`, adding a `per_page` search query to the /releases API endpoint ensured that upto 100 versions are listed, the default being 30. Go-github supports the `per_page` input through the [ListOptions](https://github.com/google/go-github/blob/master/github/github.go#L231) parameter to the ListReleases call, is adding that a valid solution?

**Comment by abhay-krishna** [comment]
Yeah that would indeed be best. In most paginated API calls, one of the fields in the response is a token that when passed into the call, fetches the next set of items. And the calls are made to the API until the token field is either empty or same as the previous call. This is prevalent in several production APIs such as those provided by AWS. I don't think GitHub does that, so what we can do is go page by page, appending the results to a list, until the results array is an empty list.

**Comment by abhay-krishna** [comment]
Was able to test with this code snippet
```go
package main

import (
	"fmt"
	"context"
	"golang.org/x/oauth2"
	"os"

	"github.com/google/go-github/v48/github"
)

func main() {
	httpTokenSource := oauth2.StaticTokenSource(
		&oauth2.Token{AccessToken: os.Getenv("GITHUB_TOKEN")},
	)
	httpClient := oauth2.NewClient(context.TODO(), httpTokenSource)
	client := github.NewClient(httpClient)
	allReleases := []*github.RepositoryRelease{}
	page := 1

	for {
		releases, _, err := client.Repositories.ListReleases(context.TODO(), "kubernetes-sigs", "cluster-api", &github.ListOptions{Page: page})
		if err != nil {
			fmt.Printf("Error listing github releases: %v\n", err)
			os.Exit(1)
		}
		if len(releases) == 0 {
			break
		}
		allReleases = append(allReleases, releases...)
		page += 1
	}

	fmt.Println(len(allReleases))
}
```

**Comment by abhay-krishna** [comment]
@killianmuldoon Opened a PR, let me know what you think!

**Comment by killianmuldoon** [comment]
Just to add some clarity and context: This is only impactful when the latest version of a release series is not in the first 30 results in the github releases. Currently the returned releases are all from the `v1beta1` release series. This would only impact a `v1beta1` release after 30 `v1beta2` releases. This means versions like `1.0.0` are available today even though it was released before `v0.4.8`
